sw/device/lib/base/mmio.c

```diff
@@ -4,6 +4,117 @@
+  if (misalignment != 0) {
+    // The number of bytes missing to bring `offset` back into alignment.
+    // For example, 0x3 has misalignment of 3 and realignment of 1.
+    ptrdiff_t realignment = sizeof(uint32_t) - misalignment;
+    // Note that we might be doing less I/O than the misalignment requires; we
+    // might be off by a single byte, but not have the full three bytes for full
+    // realignment.
+    if (realignment > len) {
+      realignment = len;
+    }
+
+    // Converts `offset`, which points to a subword boundary, to point to the
+    // start of the current word it points into.
+    ptrdiff_t current_word_offset = offset - misalignment;
+    uint32_t current_word = mmio_region_read32(base, current_word_offset);
+
+    // Act on only to a suffix of `current_word`, corresponding to the necessary
+    // realignment.
+    uint8_t *current_byte = ((uint8_t *)&current_word) + misalignment;
+    if (from_mmio) {
+      memcpy(buf, current_byte, realignment);
+    } else {
+      // When writing, we need to write the modified word.
+      memcpy(current_byte, buf, realignment);
+      mmio_region_write32(base, current_word_offset, current_word);
+    }
+
+    offset += realignment;
+    buf += realignment;
+    len -= realignment;
+  }
+
+  // Now, we just do full word I/O until we run out of stuff to act on.
+  while (len > 0) {
+    // At the end, we may not have a full word to copy, but it's otherwise
+    // the same case as a full word, since we're already word aligned (if
+    // this would be a subword read, it would end the loop anyway).
+    uint32_t bytes_to_copy = sizeof(uint32_t);
+    if (bytes_to_copy > len) {
+      bytes_to_copy = len;
+    }
+
+    // Read the current word from MMIO.
+    uint32_t current_word = 0;
+    if (from_mmio || bytes_to_copy != sizeof(uint32_t)) {
+      // If reading from MMIO, we need to read this word always.
+      // If writing to MMIO, we only need to write a prefix when writing a
+      // subword. In that case, we need to avoid clobbering the word at
+      // `offset`.
+      current_word = mmio_region_read32(base, offset);
+    }
+
+    // Copy a prefix; most of the time, this will be the whole word.
+    if (from_mmio) {
+      memcpy(buf, &current_word, bytes_to_copy);
+    } else {
+      // When writing to MMIO, we need to write the modified word.
+      memcpy(&current_word, buf, bytes_to_copy);
+      mmio_region_write32(base, offset, current_word);
+    }
+
+    offset += bytes_to_copy;
+    buf += bytes_to_copy;
+    len -= bytes_to_copy;
+  }
+}
+
+void mmio_region_memcpy_from_mmio32(mmio_region_t base, uint32_t offset,
+                                    void *dest, size_t len) {
+  mmio_region_memcpy32(base, offset, dest, len, true);
+}
+
+void mmio_region_memcpy_to_mmio32(mmio_region_t base, uint32_t offset,
+                                  const void *src, size_t len) {
+  mmio_region_memcpy32(base, offset, (void *)src, len, false);
```

Comment:
As someone reading this, I would be mildly sketched out by casing away const. Maybe note that the `false` argument means that `src` will never be written to?

Reply:
Done.